[WIP] kube-proxy: reuse node informer to improve efficiency

aojea opened the pull request:
/kind bug

Kube-proxy needs to obtain the node IPs and, in some modes, the PodCIDR value of the node object.

For the former it performs several Get requests to the apiserver, and if is not able to get in time, it keeps going using the localhost address. This PR changes to wait until the node object is present, this is a safe assumption as the kube-proxy needs to work connected to the apiserver.

For the later, it started a watch to get the Node object again, after getting the Node IPs

```release-note
kube-proxy improves the startup process by only performing at List and Watch at startup to obtain the Node information. It will block the startup process until the Node object is obtained.
```

/sig network
ping @linxiulei 

Alternative to https://github.com/kubernetes/kubernetes/pull/124670

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -628,7 +638,49 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
+	if !cache.WaitForNamedCacheSync("kube-proxy informers", ctx.Done(),
+		s.nodeInformer.Informer().HasSynced,
+		s.serviceInformer.Informer().HasSynced,
+		s.endpointSliceInformer.Informer().HasSynced) {
```
this is missing the feature fate for serviceCIDRInformer

aojea commented on cmd/kube-proxy/app/server_linux.go:
```diff
@@ -418,48 +413,25 @@ func getConntrackMax(ctx context.Context, config proxyconfigapi.KubeProxyConntra
-			return false, fmt.Errorf("event object not of type Node")
-		}
-		// don't consider the node if is going to be deleted and keep waiting
-		if !n.DeletionTimestamp.IsZero() {
```
we need to recover this condition, as we had a bug because of this, and check the node object is not being deleted

aojea commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -648,14 +649,17 @@ func Test_getNodeIPs(t *testing.T) {
 		go func() {
 			_, ctx := ktesting.NewTestContext(t)
-			ips := getNodeIPs(ctx, client, nodeName)
+			ips := getNodeIPs(ctx, nodeLister, nodeName)
```
the tests have to be updated in this file to match the code changes

linxiulei commented on cmd/kube-proxy/app/server.go:
```diff
@@ -950,52 +1002,22 @@ func (s *ProxyServer) Run(ctx context.Context) error {
-	// are registered yet.
-	serviceConfig := config.NewServiceConfig(ctx, informerFactory.Core().V1().Services(), s.Config.ConfigSyncPeriod.Duration)
+	serviceConfig := proxyconfig.NewServiceConfig(ctx, s.serviceInformer, s.Config.ConfigSyncPeriod.Duration)
 	serviceConfig.RegisterEventHandler(s.Proxier)
```
Would this be an issue since the informer starts before registration? For example, the first few events would be skipped for handlers here, right?

aojea replied:
kube-proxy reconciles, that will be a problem for event based controllers, see how informer works when the handlers are added after

https://github.com/kubernetes/kubernetes/blob/6b73cccb07de30c7793370e1d5c3f8766d975d20/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L105-L126

linxiulei commented on cmd/kube-proxy/app/server.go:
```diff
@@ -628,7 +638,49 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
+		return nil, fmt.Errorf("kube-proxy informers could not start")
+	}
+
+	rawNodeIPs := getNodeIPs(ctx, s.nodeInformer.Lister(), s.Hostname)
```
Does informer guarantee the sync-ed data is up to update with etcd rather than just API server cache?

aojea replied:
informer goes to watch cache, there is an event handler that runs later that detects if the node is stale and restart the apiserver as the Get() can also present problems wtih spot vms per example

aojea commented:
something I'm doing wrong that the informers are not starting correctly, will circle back once I have time, but this will make the startup process more reliable and more efficient

aojea committed:
replace config variable to avoid shadow with imported module

aojea committed:
kube-proxy: reuse node informer

Instead of using Get request and create a one time watcher, reuse
the existing informer starting it at the beginning, also wait
to proceed until the node object with the IP addresses are present.

aojea commented:
ok, this is working fine now

```
2024-07-01T18:42:03.813513826Z stderr F I0701 18:42:03.813385       1 feature_gate.go:385] feature gates: {map[]}
2024-07-01T18:42:03.92265811Z stderr F I0701 18:42:03.921757       1 shared_informer.go:313] Waiting for caches to sync for node informer cache
2024-07-01T18:42:03.922689959Z stderr F I0701 18:42:03.922063       1 reflector.go:305] Starting reflector *v1.Node (15m0s) from k8s.io/client-go/informers/factory.go:160
2024-07-01T18:42:03.922695602Z stderr F I0701 18:42:03.922128       1 reflector.go:341] Listing and watching *v1.Node from k8s.io/client-go/informers/factory.go:160
2024-07-01T18:42:03.931259739Z stderr F I0701 18:42:03.931118       1 reflector.go:368] Caches populated for *v1.Node from k8s.io/client-go/informers/factory.go:160
2024-07-01T18:42:04.022479372Z stderr F I0701 18:42:04.022286       1 shared_informer.go:320] Caches are synced for node informer cache
2024-07-01T18:42:04.02250784Z stderr F I0701 18:42:04.022382       1 server.go:1119] "Successfully retrieved node IP(s)" IPs=["172.18.0.3"]
```

I need to fix the unit tests

aojea committed:
remove comments that are not true, handlers can be added after the informer has started

aojea commented:
/assign @wojtek-t @thockin @linxiulei @danwinship

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -628,7 +632,24 @@ func newProxyServer(ctx context.Context, cfg *kubeproxyconfig.KubeProxyConfigura
+	nodeLister := s.nodeInformer.Lister()
+	nodeInformerHasSynced := s.nodeInformer.Informer().HasSynced
+
+	// This has to start after the calls to NewNodeConfig because that must
```
this comment is removed in the next commit

danwinship replied:
can you squash the commits together logically?

wojtek-t commented on cmd/kube-proxy/app/server.go:
```diff
@@ -1086,18 +1095,12 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
-	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
-		node, err := client.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
+	err := wait.PollUntilContextCancel(ctx, 1*time.Second, true, func(context.Context) (bool, error) {
+		node, err := nodeLister.Get(name)
```
This change means that we're switching from consistent get, to potentially stale read.
[We will fix that hopefully in 1-2 releases with ConsistentReadFromCache and/or WatchList graduation to GA, but that's not there yet].

I'm not well positioned if that will cause issues or not, but if not, I would like to see some comment explaining why.

linxiulei replied:
@aojea mentioned 

> informer goes to watch cache, there is an event handler that runs later that detects if the node is stale and restart the apiserver as the Get() can also present problems wtih spot vms per example

I might miss it but only see the check CIDR for staleness

aojea replied:
I need to revisit this logic, the node IPs seems to be used only for the primary ip family and for the loadbalancer source ranges , that now we know some times may be not working as expected as it was falling back to 127.0.0.1

/hold

linxiulei replied:
Can we just add the check with node IPs, which will make proxy restart if unmatched and we can proceed this PR?

aojea replied:
/hold cancel

added as independent check so we can remove it once the features wojtek mentions are fixed

danwinship replied:
It doesn't matter if we get stale data here; kube-proxy will exit if the node object changes later.
(And we don't expect the node IPs to change often anyway.)

wojtek-t commented on cmd/kube-proxy/app/server_linux.go:
```diff
@@ -416,48 +411,33 @@ func getConntrackMax(ctx context.Context, config proxyconfigapi.KubeProxyConntra
+	var node *v1.Node
+	err := wait.PollUntilContextCancel(ctx, 1*time.Second, true, func(context.Context) (bool, error) {
+		var listerErr error
+		node, listerErr = nodeLister.Get(nodeName)
```
This is great - as we don't regress compared to UntilWithSync.

aojea committed:
kube-proxy: restart if nodeIPs changes due to stale reads

new handler nodeIPs

wojtek-t commented on pkg/proxy/node.go:
```diff
@@ -90,6 +92,79 @@ func (n *NodePodCIDRHandler) OnNodeDelete(node *v1.Node) {
+
+// OnNodeAdd is a handler for Node creates.
+func (n *NodeIPsHandler) OnNodeAdd(node *v1.Node) {
+	n.mu.Lock()
```
Why do we need mutex? The informer guarantees that there will be at most 1 handler called at a given time.

aojea replied:
I see, I must be carrying on with some assumption of legacy code

wojtek-t commented on pkg/proxy/node.go:
```diff
@@ -90,6 +92,79 @@ func (n *NodePodCIDRHandler) OnNodeDelete(node *v1.Node) {
+// NodePodCIDRHandler handles the life cycle of kube-proxy based on the node PodCIDR assigned
+// Implements the config.NodeHandler interface
+// https://issues.k8s.io/111321
+type NodeIPsHandler struct {
```
This doesn't seem to be used anywhere in the production code...

wojtek-t replied:
OK - so it was supposed to be used but the name doesn't match - and it's what causing compile errors...

aojea replied:
yeah, heh, I have a problem that the hollow proxy fails to be initialized now

aojea committed:
kube-proxy: export NodeInformer

This is required to be able to initialize the fake Proxy used
in kubemark

aojea commented:
/retest

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -1086,22 +1098,21 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
+		if !node.DeletionTimestamp.IsZero() {
+			logger.Info("Node is being deleted", "node", node.Name)
+			return false, nil
+		}
```
Thinking about the big picture of kube-proxy Node handling, we don't really deal well with the case where the Node is deleted and then not created (but kube-proxy is still running); we will crashloop in that case. Perhaps it would be better for kube-proxy to ignore Node deletion and _only_ restart on Node recreation? (In which case, we can just ignore DeletionTimestamp)

aojea replied:
we have very strong requirements for node startup and readiness, if we wait for crashing the delay will be greater than waiting here ..

danwinship replied:
OK, but do we have strong requirements for "the Node object is deleted and then immediately recreated"?

It seems weird to have kube-proxy have to specifically distinguish the case of "the Node is in the process of being deleted but hasn't quite been deleted yet". Someone's abstraction is leaking all over the place.

aojea replied:
It's an edge case that happens because of two factor:
1. The VM is quickly terminated and recreated O(seconds):  Spot VM, preempted VMs or kind of VM migration situation (not live of course as the system is restarted). Common in cloud providers, at least common in GCE but afaik also AWS does something similar 
2. Run kube-proxy as a static pod, instead of a daemon set, so you kproxy on the node is tied to the node version 

I don't know now from the top of my mind of this deletion timestamp was me being paranoid or part of the bug, but with this setup you'll face the race eventually and your kproxy will get values from the old node object

danwinship replied:
> 1. The VM is quickly terminated and recreated

With the Node object being deleted or not?

- **Node is deleted, then VM is restarted**: This is no different from the initial Node creation case; the VM will come up, kubelet and kube-proxy will be started, and kube-proxy has to wait for kubelet to create a new Node object. (So this already works fine.)
- **Node not deleted, Node not updated after VM restart**: Again, this already works fine; the correct Node will already exist when kube-proxy restarts.
- **Node not deleted, Kubelet updates it after VM restart**: (eg, because the VM gets a different IP after rebooting). In this case, it's possible that kube-proxy will start up and read the "old" Node, and then eventually kubelet will update the Node object and kube-proxy will have to restart (at which point it basically becomes the previous case: when kube-proxy restarts, the correct Node object will already exist).
- **VM restarted, then Node deleted**: In this case, kube-proxy will (potentially) start up with the "old" Node, as above, until it notices the Node deletion and restarts. If we make it so that (a) kube-proxy exits as soon as it sees a DeletionTimestamp on the Node rather than waiting for the deletion event, and (b) kube-proxy ignores a Node with a DeletionTimestamp at startup time, then we potentially get kube-proxy back up and running with correct configuration faster than we would have if we didn't do both (a) and (b).

So, yes, handling DeletionTimestamp can help in one case, but what would help even more would be if the infra made sure that if there's any chance the Node will need to be deleted or updated after the VM restart, then it should delete the Node before restarting...

> Run kube-proxy as a static pod, instead of a daemon set, so you kproxy on the node is tied to the node version

But what are the scenarios with that that would require the DeletionTimestamp check?

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -1086,22 +1098,21 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
+		}
 		nodeIPs, err = utilnode.GetNodeHostIPs(node)
 		if err != nil {
 			logger.Error(err, "Failed to retrieve node IPs")
```
This should be `Info` (though with the `err` included). It most likely just indicates that we fetched the Node in between when kubelet created it and when the cloud provider filled in the IPs.

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -1111,6 +1122,8 @@ func getNodeIPs(ctx context.Context, client clientset.Interface, name string) []
 	if err == nil {
 		logger.Info("Successfully retrieved node IP(s)", "IPs", nodeIPs)
+	} else {
+		logger.Info("Error to retrieve node IP(s)", "error", err)
```
Whereas this is actually an error. But `err` is just going to be `timed out`.
It might be nicer to do the "preserve the last error returned from GetNodeHostIPs and only log it if the loop eventually times out rather than logging it each time" thing.

danwinship commented on cmd/kube-proxy/app/server_linux.go:
```diff
@@ -416,48 +411,33 @@ func getConntrackMax(ctx context.Context, config proxyconfigapi.KubeProxyConntra
-		// don't process delete events
-		if event.Type != watch.Modified && event.Type != watch.Added {
+	var node *v1.Node
+	err := wait.PollUntilContextCancel(ctx, 1*time.Second, true, func(context.Context) (bool, error) {
```
We could merge this with the other loop, and wait for both bits of data in one place (in the NodeCIDR-local-detector case)

aojea replied:
yeah, all these logic is very spaguetti, I need to find a some time to come up with a more consistent refactor that allow us to improve current code

danwinship commented on cmd/kube-proxy/app/server_linux_test.go:
```diff
@@ -493,51 +495,56 @@ detectLocalMode: "BridgeInterface"`)
-		if err != nil {
-			t.Fatalf("unexpected error when setting up environment: %v", err)
-		}
+		t.Run(tc.name, func(t *testing.T) {
```
(could have been a separate commit to clarify that it isn't making any changes)

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -1009,7 +1010,13 @@ func (s *ProxyServer) Run(ctx context.Context) error {
+	// The solution in the meantime is to process the Node events and crash if the NodeIPs or the
+	// the PodCIDRs (if local mode nodeCIDR is used) has changed since the first read.
 	// https://issues.k8s.io/111321
+	nodeConfig.RegisterEventHandler(proxy.NewNodeIPsHandler(ctx, s.rawNodeIPs))
```
yeah, we we should _really_ just merge all the node watching handling together. It's silly to have separate node IPs and node podCIDRs watchers.

danwinship commented on cmd/kube-proxy/app/server_linux_test.go:
```diff
@@ -516,7 +516,10 @@ detectLocalMode: "BridgeInterface"`)
+				_, err = file.WriteString("append fake content")
+				if err != nil {
+					t.Fatal(err)
+				}
```
squash into previous commit

danwinship commented on pkg/proxy/kubemark/hollow_proxy.go:
```diff
@@ -63,6 +66,12 @@ func NewHollowProxy(
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", nodeName).String()
+		}))
+	currentNodeInformerFactory.Start(wait.NeverStop)
```
HollowProxy shouldn't have to know this much about `cmd/kube-proxy/`'s internals, if possible

aojea commented:
/test

aojea commented:
@adrianmoisey 👀 

/test pull-kubernetes-e2e-gce-network-policies

aojea commented:
/test pull-kubernetes-e2e-gce-network-policies

aojea commented:
/test pull-kubernetes-e2e-gce-network-policies

it didn't merge yet

aojea commented:
/test pull-kubernetes-e2e-kind-dual-canary

aojea commented:
/test pull-kubernetes-e2e-kind-nftables

adrianmoisey commented:
/test pull-kubernetes-e2e-gce-network-policies

/test pull-kubernetes-e2e-kind-dual-canary

aojea commented:
/test pull-kubernetes-e2e-gce-network-policies

/test pull-kubernetes-e2e-kind-dual-canary

aojea commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer

aojea commented:
/test pull-kubernetes-e2e-gce-network-policies

aojea commented:
/test pull-kubernetes-e2e-gce-network-policies

/test pull-kubernetes-e2e-kind-dual-canary

aojea commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer

danwinship reviewed (commented):
(some earlier review comments still apply too)

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -196,7 +200,24 @@ func newProxyServer(ctx context.Context, cfg *kubeproxyconfig.KubeProxyConfigura
+	// configure the shared informer event handler first.
+	currentNodeInformerFactory.Start(wait.NeverStop)
+
+	if !cache.WaitForNamedCacheSync("node informer cache", ctx.Done(), nodeInformerHasSynced) {
```
independently , if is a noop there is no problem, and if some future us depend on the cache to be sync, we avoid a possible problem

aojea commented:
I need to retake this, no time at this moment, weill remove the WIP when is ready to review again
